src/API.jl

```diff
@@ -895,30 +895,72 @@ end
+    pkg_dep_uuid_lists = [collect(values(last(dep).deps)) for dep in man if !Pkg.Operations.is_stdlib(first(dep))]
+    filter!.(!is_stdlib, pkg_dep_uuid_lists)
+    
     if ctx.env.pkg !== nothing && isfile( joinpath( dirname(ctx.env.project_file), "src", ctx.env.pkg.name * ".jl") )
         push!(pkgids, Base.PkgId(ctx.env.pkg.uuid, ctx.env.pkg.name))
+        push!(pkg_dep_uuid_lists, collect(keys(ctx.env.project.deps)))
+    end    
+    
+    was_processed = Dict{Base.UUID,Base.Event}()
+    was_recompiled = Dict{Base.UUID,Bool}()
+    for pkgid in pkgids
+        was_processed[pkgid.uuid] = Base.Event()
+        was_recompiled[pkgid.uuid] = false
     end
-
-    # TODO: since we are a complete list, but not topologically sorted, handling of recursion will be completely at random
-    for pkg in pkgids
+    
+    function is_stale(paths, sourcepath)
+        for path_to_try in paths::Vector{String}
+            staledeps = Base.stale_cachefile(sourcepath, path_to_try, Base.TOMLCache())
+            staledeps === true && continue
+            return false
+        end
+        return true
+    end
+    
+    errored = false
+    @sync for (i, pkg) in pairs(pkgids)
         paths = Base.find_all_in_cache_path(pkg)
         sourcepath = Base.locate_package(pkg)
         sourcepath === nothing && continue
         # Heuristic for when precompilation is disabled
         occursin(r"\b__precompile__\(\s*false\s*\)", read(sourcepath, String)) && continue
-        stale = true
-        for path_to_try in paths::Vector{String}
-            staledeps = Base.stale_cachefile(sourcepath, path_to_try)
-            staledeps === true && continue
-            # TODO: else, this returns a list of packages that may be loaded to make this valid (the topological list)
-            stale = false
-            break
-        end
-        if stale
-            Base.compilecache(pkg, sourcepath)
-        end
+        
+        @async begin
+            for dep_uuid in pkg_dep_uuid_lists[i] # wait for deps to finish
+                wait(was_processed[dep_uuid])
+            end
+            
+            # skip stale checking and force compilation if any dep was recompiled in this session
+            any_dep_recompiled = any(map(dep_uuid->was_recompiled[dep_uuid], pkg_dep_uuid_lists[i]))
+            if !errored && (any_dep_recompiled || is_stale(paths, sourcepath))
+                Base.acquire(parallel_limiter)
+                if errored # catch things queued before error occurred
+                    notify(was_processed[pkg.uuid])
+                    Base.release(parallel_limiter)
+                    return
+                end
+                try
+                    Base.compilecache(pkg, sourcepath)
+                    was_recompiled[pkg.uuid] = true
+                catch err
+                    errored = true
+                    throw(err)
```

Comment:
I couldn't bring https://github.com/JuliaLang/julia/pull/37374#issuecomment-689256691 up again in this PR in time, but did somebody considered the case for OS-specific packages? That is to say, if we have

```julia
module DownstreamPkg
if Sys.iswindows()
    using SomeWindowsOnlyPackage
end
...
end
```

we don't need to compile `SomeWindowsOnlyPackage` (or propagate the error from compiling `SomeWindowsOnlyPackage` in non-Windows OS) when compiling `DownstreamPkg` in non-Windows OS.

I think we should report errors only from the importable packages (i.e., the ones in Project.toml). Of course, it'd be much nicer to integrate precompilation machinery in Base itself so that this kind of hack is not necessary.

Reply:
Is that what makes it hard to parallelize in Base, because you don't know your actual dependencies up front?

Reply:
I see. Perhaps the docstring for `Pkg.precompile` could be changed to

```
help?> Pkg.precompile
  Pkg.precompile()

  Parallelized precompilation of all the dependencies in the manifest of the project. 
  If you want to only precompile the packages that are actually used, instead load the packages to trigger standard precompilation.
...
```

Reply:
I'll let others decide if this is necessary, but it proved simple to implement an option to revert behavior https://github.com/JuliaLang/Pkg.jl/pull/2021

Reply:
> Is that what makes it hard to parallelize in Base, because you don't know your actual dependencies up front?

Yeah.  But it's a problem in Pkg, too (even if you can parse Manifest).  I think the problem is that there is no database for "true" dependency tree since you can hide `using` in `if` branch.

Ultimately, only the precompilation subprocess "knows" the dependencies while it's precompiling a package. So, I think the "correct" way of doing this to use some kind of bidirectional RPC so that the precompilation subprocess can request the precompilation of its dependencies to the parent process. But that's very tedious. I think ignoring compilation error from non-direct dependencies is a decent approach.

(I feel like I should try implementing it now that I complained so much :rofl:)

Reply:
Neat idea on non-direct dependencies being allowed to fail.. I implemented it here https://github.com/JuliaLang/Pkg.jl/pull/2021#issuecomment-693095161

Reply:
Wouldn't it be enought to wrap in 
```
redirect_stderr(devnull)
```
here? That should propagate to the subprocess, no?

Reply:
That's implemented in https://github.com/JuliaLang/Pkg.jl/pull/2021 but there was a question about `redirect_stderr` concurrency here https://github.com/JuliaLang/Pkg.jl/pull/2021#discussion_r489098311 so https://github.com/JuliaLang/julia/pull/37596 was opened